### ISSUE TITLE
Don't write .ep_initialized in source directories

### ISSUE THREAD

**Issue by mitar** [ether/etherpad-lite#2683]
In my installation I install Etherpad without giving write permissions to the node processor over the source directory. It seems this is a problem because Etherpad tries to write to four places:
- `APIKEY.txt` (OK, I can generate this in advance)
- `SESSIONKEY.txt` (I can generate this as well)
- `var/` (this is easy to make writable by node process)
- `.ep_initialized` in each plugin directory – this is hard to do without giving write access to all plugin code

So why is `.ep_initialized` needed? It seems it is only created once, but never really cleaned up?

**Comment by JohnMcLear** [comment]
It let's Etherpad know a plugin is initialised.  I can't remember if we still need it or not

**Comment by mitar** [comment]
I solved this by simply creating the files in advance. I do not see any downsides.

**Comment by mitar** [comment]
What is decision here? Why just closing it? This is a real problem for properly Dockerizing the Etherpad without giving write access to all files: https://github.com/tozd/docker-etherpad

**Comment by JohnMcLear** [comment]
Well you solved and documented the fix, others can follow suit.  If you want a more perma fix I'd be happy to see a pull request with it in.

**Comment by mitar** [comment]
It is unclear if this works correctly. Because if it works correctly, then we can just remove all this code. I can create the pull request for that.

**Comment by JohnMcLear** [comment]
This is the source code that references the file
```
      fs.stat(path.normalize(path.join(plugin.package.path, ".ep_initialized")), function (err, stats) {
        if (err) {
          async.waterfall([
            function (cb) { fs.writeFile(path.normalize(path.join(plugin.package.path, ".ep_initialized")), 'done', cb); },
            function (cb) { hooks.aCallAll("init_" + plugin_name, {}, cb); },
            cb,
          ]);
        } else {
          cb();
        }
      });
```

**Comment by JohnMcLear** [comment]
In some plugins you can make it do a "one time execution" such as a database preparation.  It appears this does this.

**Comment by mitar** [comment]
I have not seen any plugin defining `"init_" + plugin_name`. Also, probably it is better for a plugin to check themselves if a database is missing and then initialize. Such file based approach is pretty fragile.

I would vote to remove it. (Is this even documented anywhere?)

**Comment by JohnMcLear** [comment]
*sighs
http://etherpad.org/doc/v1.5.6/#index_init_plugin_name

**Comment by mitar** [comment]
OK. But then this ticket is not resolved. Having to write to the file system to mark makes it hard to lock down. My workaround might break real plugins. So I think this should be reopened. Maybe the list of initialized plugins could be stored in the database?

**Comment by JohnMcLear** [comment]
If your container can't handle writing to the filesystem I just don't see this as being an Etherpad problem however I would take a ``clean`` solution that used the database instead.  

You might need to be mindful that the database might not be available/ready and/or might be shared/sharded etc. so this might be a really bad solution for what you are proposing.

TLDR; Not Etherpad problem, container problem.

**Comment by mitar** [comment]
OK, what about this, instead of writing those files into the source directory, we should write them into the `var/` directory? So something like `var/plugin_{plugin_name}_initialized` file?

So it is easy to make the whole (source-less) directory writable by the Etherpad node.js process. But one cannot make it easy to create only one file, but prevent writing to other source files of the module.

**Comment by JohnMcLear** [comment]
Yea that might be a clean enough solution, give it a go and see if it works.  You will need to create a test plugin to ensure it does.

**Comment by LaKing** [comment]
I also experience this as a bit problematic, the sourced folders should not be writable.

**Comment by mitar** [comment]
unstale

**Comment by mitar** [comment]
unstale

**Comment by mitar** [comment]
unstale

**Comment by mitar** [comment]
unstale

**Comment by JohnMcLear** [comment]
Are you going to work on this @mitar ?

**Comment by mitar** [comment]
No time at the moment, but eventually yes, if somebody else does not do it earlier.

**Comment by JohnMcLear** [comment]
Just looking at this, it's easy enough to just write toe database instead of file system...  Wouldn't be a hard fix at all... 

``db.set("someKey:initialized", true)``; to set

``await db.get("someKey:initialized")``; to get

no need to use file system if you do that..  One makes assumption database is init'd by this point tho?

**Comment by rhansen** [comment]
Why is `.ep_initialized` written in the first place?

**Comment by JohnMcLear** [comment]
*gets out his time machine..

So basically when the plugin framework was first designed there was a method that can be executed by a plugin if it was the first run.  Afaik it was for creaitng a table in a database for a plugin

https://github.com/redhog/ep_sketchspace/blob/master/ep.json#L6

I was wrong, it was to run a ``build.sh`` script for this plugin.

ep_mypads uses it https://github.com/framasoft/ep_mypads/blob/master/ep.json#L7 - https://github.com/framasoft/ep_mypads/blob/master/hooks.js#L40

We didn't switch to writing to database for the file because a database isn't necessarily a truthful representation of a file system IE you might have a cluster of etherpad frontend servers and only one backend..
